### PR TITLE
Use 2's complement when computing the relative discriminant to avoid signed to unsigned overflow failures

### LINKED ISSUE #356: Verification fails when comparing arrays and string slices.
I tried this code:

```rust
fn main() {
    assert!([1, 2, 3] < [1, 3, 4]);
    assert!("World" >= "Hello");
}
```

using the following command line invocation:

```
rmc test.rs
```

with RMC version: 6962a6700600c98a665a8547739b743f490c15d9

I expected to see this happen: verification succeeds.

Instead, this happened: verification failed with the following assertions failures:
```
[_RNvYSlNtNtCsgWci0eQkB8o_4core3cmp10PartialOrd2ltCs8ZdcvQGSn7l_4test.overflow.1] line 1018 arithmetic overflow on signed - in *((signed char *)((unsigned char *)&var_3 + 0)) - 2: FAILURE
[_RNvYSlNtNtCsgWci0eQkB8o_4core3cmp10PartialOrd2ltCs8ZdcvQGSn7l_4test.overflow.2] line 1018 arithmetic overflow on signed to unsigned type conversion in (unsigned long int)(*((signed char *)((unsigned char *)&var_3 + 0)) - 2): FAILURE
[_RNvYeNtNtCsgWci0eQkB8o_4core3cmp10PartialOrd2geCs8ZdcvQGSn7l_4test.overflow.2] line 1078 arithmetic overflow on signed to unsigned type conversion in (unsigned long int)(*((signed char *)((unsigned char *)&var_3 + 0)) - 2): FAILURE
[main.assertion.1] line 2 assertion failed: [1, 2, 3] < [1, 3, 4]: FAILURE
```

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

An overflow may occur when casting a signed value to an unsigned one in the computation of an enum discriminant. This overflow is intentional, but gets flagged by CBMC as a failure. This PR works around the issue by using the 2's complement to avoid overflow.

### Resolved issues:

Resolves #356 

### Call-outs:

An alternative approach would be to use CBMC pragmas to turn off signed to unsigned overflow detection during the computation of the relative discriminant. We haven't had much success with the usage of such pragmas in the past though.

### Testing:

* How is this change tested? Two new tests are added.

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X]  My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 1]
```diff
@@ -467,8 +493,8 @@ impl<'tcx> GotocCtx<'tcx> {
                     } else {
                         relative_discr
                             .clone()
-                            .cast_to(Type::unsigned_int(64))
-                            .le(Expr::int_constant(relative_max, Type::unsigned_int(64)))
+                            .cast_to(Type::unsigned_int(128))
```
Comment (celinval) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:496]: Why do you need this?

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 2]
```diff
@@ -457,8 +459,32 @@ impl<'tcx> GotocCtx<'tcx> {
                     let relative_discr = if *niche_start == 0 {
                         niche_val
                     } else {
-                        // This should be a wrapping sub.
-                        niche_val.sub(Expr::int_constant(*niche_start, discr_ty.clone()))
+                        // Compute "niche_val - niche_start" where "-" is
```
Comment (celinval) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:462]: nit: can you move this to a different function to make this more readable

File: tests/kani/Enum/neg_discriminant.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+#[kani::proof]
+fn main() {
+    let a = Some(Foo::A);
+    let b = Some(Foo::B);
+    let c = Some(Foo::C);
+    let d = Some(Foo::D);
+    let e = Some(Foo::E);
+    let f = Some(Foo::F);
+    let _ = matches!(a, Some(Foo::A));
```
Comment (celinval) [inline tests/kani/Enum/neg_discriminant.rs:23]: can we assert?

File: tests/kani/Enum/neg_discriminant.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    A = -500,
+    B = -200,
+    C = -100,
+    D = 0,
+    E = 1,
+    F = 256,
+}
+
+#[kani::proof]
+fn main() {
```
Comment (celinval) [inline tests/kani/Enum/neg_discriminant.rs:16]: nit: I think we should stop using main() and add a more descriptive name to the harness now that we run them by default.

File: tests/kani/LexicographicCmp/main.rs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This test checks that lexicographic comparison is handled correctly
+
+#[kani::proof]
+fn main() {
```
Comment (celinval) [inline tests/kani/LexicographicCmp/main.rs:7]: verify_lexicographic_comparison

**Review by celinval — COMMENTED** [review]
Thanks for doing this. Mostly small comments.

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs [hunk 6]
```diff
@@ -457,8 +459,32 @@ impl<'tcx> GotocCtx<'tcx> {
...
+                        //   if niche_val >= niche_start {
+                        //       // result is positive, so overflow may not occur
+                        //       niche_val - niche_start
+                        //   } else {
+                        //       // compute the 2's complement to avoid overflow
+                        //       niche_val - niche_start + 2^32
+                        //   }
+                        let s64 = Type::signed_int(64);
+                        let niche_val = niche_val.cast_to(s64.clone());
+                        let twos_complement: i64 =
```
Comment (danielsn) [inline src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs:481]: I'm adding a general 2's complement in #981

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs (hunk 1 above)
Comment (zhassan-aws) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs]: Good point. It is not needed. I removed it and used `relative_discr`'s type for `relative_max`.

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs (hunk 2 above)
Comment (zhassan-aws) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs]: Done.

File: tests/kani/Enum/neg_discriminant.rs (hunk 3 above)
Comment (zhassan-aws) [reply tests/kani/Enum/neg_discriminant.rs]: Done.

File: tests/kani/LexicographicCmp/main.rs (hunk 5 above)
Comment (zhassan-aws) [reply tests/kani/LexicographicCmp/main.rs]: Done (used `check_lexicographic_cmp`).

File: src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs (hunk 6 above)
Comment (zhassan-aws) [reply src/kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs]: Thanks for adding a function for a general 2's complement! This one is for a specific width though, so was simple enough.

**Review by celinval — APPROVED** [review]
Looks great! Thanks

**Pull request merged**